lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +324,335 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
   QueuedTasks.push(std::move(T));
 }
 
-bool TaskQueue::execute(TaskBeganCallback Began, TaskFinishedCallback Finished,
-                        TaskSignalledCallback Signalled) {
+/// Owns Tasks, handles correspondence between Tasks, file descriptors, and
+/// process IDs.
+/// FIXME: only handles stdout pipes, ignores stderr pipes.
+class TaskMap {
   using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
+  PidToTaskMap TasksByPid;
 
-  // Stores the current executing Tasks, organized by pid.
-  PidToTaskMap ExecutingTasks;
+public:
+  TaskMap() = default;
 
-  // Maintains the current fds we're checking with poll.
-  std::vector<struct pollfd> PollFds;
+  bool empty() const { return TasksByPid.empty(); }
+  unsigned size() const { return TasksByPid.size(); }
 
-  bool SubtaskFailed = false;
+  void add(std::unique_ptr<Task> T) { TasksByPid[T->getPid()] = std::move(T); }
 
-  unsigned MaxNumberOfParallelTasks = getNumberOfParallelTasks();
+  Task &findTaskForFd(const int fd) {
+    auto predicate = [&fd](PidToTaskMap::value_type &value) -> bool {
+      return value.second->getPipe() == fd;
+    };
+    auto iter = std::find_if(TasksByPid.begin(), TasksByPid.end(), predicate);
+    assert(iter != TasksByPid.end() &&
+           "All outstanding fds must be associated with a  Task");
+    return *iter->second;
+  }
 
-  if (MaxNumberOfParallelTasks == 0)
-    MaxNumberOfParallelTasks = 1;
+  void destroyTask(Task &T) { TasksByPid.erase(T.getPid()); }
+};
 
-  while ((!QueuedTasks.empty() && !SubtaskFailed) ||
-         !ExecutingTasks.empty()) {
-    // Enqueue additional tasks, if we have additional tasks, we aren't
-    // already at the parallel limit, and no earlier subtasks have failed.
-    while (!SubtaskFailed && !QueuedTasks.empty() &&
-           ExecutingTasks.size() < MaxNumberOfParallelTasks) {
-      std::unique_ptr<Task> T(QueuedTasks.front().release());
-      QueuedTasks.pop();
-      if (T->execute())
-        return true;
+/// Concurrently execute the tasks in the TaskQueue, collecting the outputs from
+/// each task.
+/// Maintain invarients connecting tasks to execute, tasks currently executing,
+/// and fds being polled. These invarients include:
+/// A task is not in both TasksToBeExecuted and TasksBeingExecuted,
+/// A task is executing iff it is in TasksBeingExecuted,
+/// A task is executing iff any of its fds being polled are in FdsBeingPolled
+/// (These should be all of its output fds, but today is only stdout.)
+/// When a task has finished executing, TaskHospice waits for it to die, takes
+/// action appropriate to the cause of death, then TaskMonitor reclaims its
+/// storage.
+class TaskMonitor {
+  std::queue<std::unique_ptr<Task>> &TasksToBeExecuted;
+  TaskMap TasksBeingExecuted;
+
+  std::vector<struct pollfd> FdsBeingPolled;
+
+  const unsigned MaxNumberOfParallelTasks;
 
-      pid_t Pid = T->getPid();
+public:
+  struct CallBacks {
```

Comment:
Nitpick: We seem to treat this as a single word in this codebase, "Callbacks".

Reply:
OK, tx. Fixed.